tempodb/encoding/vparquet/parquet_reader.go

```diff
@@ -0,0 +1,86 @@
+package vparquet
+
+import (
+	"context"
+	"encoding/binary"
+	"io"
+
+	"github.com/google/uuid"
+
+	"github.com/grafana/tempo/tempodb/backend"
+)
+
+type BackendReaderAt struct {
+	ctx      context.Context
+	r        backend.Reader
+	name     string
+	blockID  uuid.UUID
+	tenantID string
+
+	TotalBytesRead uint64
+}
+
+var _ io.ReaderAt = (*BackendReaderAt)(nil)
+
+func NewBackendReaderAt(ctx context.Context, r backend.Reader, name string, blockID uuid.UUID, tenantID string) *BackendReaderAt {
+	return &BackendReaderAt{ctx, r, name, blockID, tenantID, 0}
+}
+
+func (b *BackendReaderAt) ReadAt(p []byte, off int64) (int, error) {
+	b.TotalBytesRead += uint64(len(p))
+	err := b.r.ReadRange(b.ctx, b.name, b.blockID, b.tenantID, uint64(off), p, false)
+	return len(p), err
+}
+
+func (b *BackendReaderAt) ReadAtWithCache(p []byte, off int64) (int, error) {
+	err := b.r.ReadRange(b.ctx, b.name, b.blockID, b.tenantID, uint64(off), p, true)
+	return len(p), err
+}
+
+type parquetOptimizedReaderAt struct {
+	r          io.ReaderAt
+	br         *BackendReaderAt
+	readerSize int64
+	footerSize uint32
+
+	// storing offsets and length of objects we want to cache
+	cachedObjects map[int64]int64
+}
+
+var _ io.ReaderAt = (*parquetOptimizedReaderAt)(nil)
+
+// this will be called by parquet-go in OpenFile() to set offset and length of footer section
+func (r *parquetOptimizedReaderAt) SetFooterSection(offset, length int64) {
+	r.cachedObjects[offset] = length
+}
+
+// this will be called by parquet-go in OpenFile() to set offset and length of column indexes
+func (r *parquetOptimizedReaderAt) SetColumnIndexSection(offset, length int64) {
+	r.cachedObjects[offset] = length
+}
+
+// this will be called by parquet-go in OpenFile() to set offset and length of offset index section
+func (r *parquetOptimizedReaderAt) SetOffsetIndexSection(offset, length int64) {
+	r.cachedObjects[offset] = length
+}
+
+func (r *parquetOptimizedReaderAt) ReadAt(p []byte, off int64) (int, error) {
+	if len(p) == 4 && off == 0 {
+		// Magic header
+		return copy(p, []byte("PAR1")), nil
+	}
+
+	if len(p) == 8 && off == r.readerSize-8 && r.footerSize > 0 /* not present in previous block metas */ {
+		// Magic footer
+		binary.LittleEndian.PutUint32(p, r.footerSize)
+		copy(p[4:8], []byte("PAR1"))
+		return 8, nil
+	}
+
+	// check if the offset and length is stored as a special object
+	if r.cachedObjects[off] == int64(len(p)) {
```

Comment:
we're kind of "forking" the read path here. We're sending special queries off to the backend reader and other queries to the in memory buffered reader at. Is there a cleaner implementation here?

It feels like all calls should go through the same layers:
Parquet File -> In memory cache -> Off process cache (for specific segments of the file) -> Object storage

Is there a clean way to achieve this?

Reply:
All calls are going through the same layers, ie. Tempo Querier -> Cached Reader -> Backend reader -> Object storage. The same underlying `BackendReaderAt` is called each time. If you look at this piece of logic here - https://github.com/grafana/tempo/pull/1564/files#diff-6c42681fadfdec225eb9164d0d49a2fe2033bb8ff0e820ecbd5010bb0bb629d8R37-R51
The difference is that just one of the readers is buffered.

We are just setting `shouldCache` to `true` in some special cases and leaving it as `false` for most other calls. 

Not sure if I'm missing something, we can also chat offline to discuss.

Reply:
Yeah, I can't say I have great ideas here. I think what I expect in some perfect world is 

- only one "reader" being passed to this object which all calls go through. 
- the special ranges (footer, metadata, etc) also passing through the in memory buffer 

However, I think any attempt to do this while also caching the special ranges would be difficult and would make other sections of code worse. My largest concern is that by adding this code we will be increasing calls to the backend especially in cases where off process cache is not configured. Here is a read listing on searching the cluster column:

```
len=77631 off=664715997 hit=false dur=127.416656ms
len=4096 off=4645575 hit=false dur=324.308924ms
len=4096 off=4649671 hit=true dur=324.310427ms
len=4096 off=4653767 hit=true dur=324.312226ms
...
len=4096 off=47668776 hit=false dur=695.911273ms
len=4096 off=47672872 hit=true dur=695.912669ms
len=4096 off=47676968 hit=true dur=695.913873ms
...
len=4096 off=85484375 hit=false dur=1.965658619s
len=4096 off=85488471 hit=true dur=1.96566025s
len=4096 off=85492567 hit=true dur=1.965661536s
...

The first read is the footer. Then it attempts to start reading the column in one of the row groups. You can see there is a buffered cache miss (hit = false) everytime it jumps to a new row group. If that first call is grabbing some kind of section metadata would it increase the number of calls to the backend b/c it wouldn't be hitting the buffered reader?

Maybe all I'm wanting here is a way to configure this on/off to assess impact.

Reply:
Discussed offline, and went ahead with making cache settings configurable. Let me know if this works!